zjit/src/codegen.rs

```diff
@@ -1071,14 +1071,28 @@ fn gen_guard_type(jit: &mut JITState, asm: &mut Assembler, val: lir::Opnd, guard
         assert!(Qfalse.as_i64() == 0);
         asm.test(val, val);
         asm.jne(side_exit(jit, state, GuardType(guard_type))?);
+    } else if guard_type.is_immediate() {
+        // All immediate types' guard should be inlined
+        panic!("unexpected immediate guard type: {guard_type}");
     } else if let Some(expected_class) = guard_type.runtime_exact_ruby_class() {
-        asm_comment!(asm, "guard exact class");
+        asm_comment!(asm, "guard exact class for non-immediate types");
 
-        // Get the class of the value
-        let klass = asm.ccall(rb_yarv_class_of as *const u8, vec![val]);
+        let side_exit = side_exit(jit, state, GuardType(guard_type))?;
+
+        // Check if it's a special constant
+        asm.test(val, (RUBY_IMMEDIATE_MASK as u64).into());
+        asm.jnz(side_exit.clone());
+
+        // Check if it's false
+        assert!(Qfalse.as_i64() == 0);
+        asm.test(val, val);
+        asm.jz(side_exit.clone());
```

Comment:
I think it's a result of a micro-optimization in YJIT, but the backend is capable of doing that today. Without adding the assertion, you can write:

```suggestion
        asm.cmp(val, Qfalse.into());
        asm.je(side_exit.clone());
```

and the x86 backend will convert `cmp(val, 0)` to `test(val, val)`.

Reply:
Thanks. I also update the condition for false type guard.